tests/model/model.cpp

```diff
@@ -112,6 +112,212 @@ TEST(Model, add_components) {
+    libcellml::Model m;
+    libcellml::Component c1;
+    c1.setName("child1");
+    m.addComponent(c1);
+
+    libcellml::Component& cA = m.getComponent(0);
+    cA.setName("childA");
+
+    std::string a = m.serialise(libcellml::CELLML_FORMAT_XML);
+    EXPECT_EQ(e, a);
+
+    // Using const version of overloaded method
+    libcellml::Component cB = static_cast<const libcellml::Model>(m).getComponent(0);
+    // Can't do this as I have said it is a const reference
+    // cB.setName("gus");
+    EXPECT_EQ("childA", cB.getName());
+
+    EXPECT_THROW(m.getComponent(4), std::out_of_range);
+}
+
+TEST(Model, take) {
+    const std::string e = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<model xmlns=\"http://www.cellml.org/cellml/1.2#\"></model>";
+    libcellml::Model m;
+    libcellml::Component c1, c2;
+    c1.setName("child1");
+    c2.setName("child2");
+    m.addComponent(c1);
+    m.addComponent(c2);
+
+    libcellml::Component c02 = m.takeComponent(1);
+    EXPECT_EQ(1, m.componentCount());
+
+    EXPECT_THROW(m.takeComponent(4), std::out_of_range);
+
+    EXPECT_EQ("child2", c02.getName());
+
+    libcellml::Component c01 = m.takeComponent("child1");
+    EXPECT_EQ(0, m.componentCount());
+
+    EXPECT_EQ("child1", c01.getName());
+    std::string a = m.serialise(libcellml::CELLML_FORMAT_XML);
+    EXPECT_EQ(e, a);
+
+    EXPECT_THROW(m.takeComponent("child4"), std::out_of_range);
+}
+
+static int count = 0;
+class big_and_complicated {
+   // lots of complicated code
+public:
+    int id;
+    big_and_complicated()
+        : id(count+101)
+    {
+        count++;
+    }
+
+    ~big_and_complicated()
+    {
+        count--;
+    }
+};
+
+struct structure
+{
+    structure()
+      : m_data{new big_and_complicated{}}
+    {
+        std::cout << "structure constructor: " << m_data->id << std::endl;
+    }
+
+    structure(const structure& rhs)
+        : m_data{new big_and_complicated{}}
+    {
+        std::cout << "structure copy constructor: " << rhs.m_data->id << std::endl;
+        m_data->id = rhs.m_data->id;
+    }
+
+    structure(structure&& rhs)
+      : m_data(rhs.m_data)
+    {
+        std::cout << "structure move constructor: " << m_data->id << std::endl;
+        rhs.m_data = nullptr;
+    }
+
+    structure &operator=(structure r)
+    {
+        r.swap(*this);
+        return *this;
+    }
+
+    void swap(structure &r) throw()
+    {
+        std::swap(this->m_data, r.m_data);
+    }
+
+    ~structure()
+    {
+        std::cout << "structure destructor: ";
+        if (m_data) {
+            std::cout << m_data->id << std::endl;
+        } else {
+            std::cout << std::endl;
+        }
+        delete m_data;
+    }
+
+private:
+    big_and_complicated* m_data;
+};
+
+//TEST(Model, simplified_replace) {
+//    std::vector<structure> ss;
+//    structure s1, s2, s3;
+//    ss.push_back(s1);
+//    ss.push_back(s2);
+//    ss.erase(ss.begin());
+//    ss.insert(ss.begin(), s3);
+
+//}
+
```

Comment:
Shouldn't all that code for ```big_and_complicated```, ```structure``` and ```TEST(Model, simplified_replace) ``` be removed since it's not being used and doesn't seem to be related to the issue at hand?